### PR TITLE
Uptimize search of disruptions for a physical_mode

### PR DESCRIPTION
**Pull request by kadhikari**
Tested with gtfs-rt from kirin on coverage sncf of dev -> compared between dev and in local with the same data and gtfs-rt
For details refer to https://jira.kisio.org/browse/NAVITIAII-3622

### CODE REVIEW

**Comment by xlqian** [comment]
why not add a shortcut in the data graph?

File: source/ptreferential/ptreferential_utils.cpp [hunk 1]
```diff
@@ -69,6 +69,24 @@ Indexes get_intersection(const Indexes& idxs1, const Indexes& idxs2) {
...
+        const auto vjs = data.get_data<type::VehicleJourney>(vj_idxs);
+        Indexes vj_indexes, temp;
+        for (type::VehicleJourney* vj : vjs) {
+            if (indexes.find(vj->physical_mode->idx) != indexes.cend()) {
+                // Add Impact on vj
+                temp = vj->get(Type_e::Impact, *(data.pt_data.get()));
+                vj_indexes.insert(temp.begin(), temp.end());
+            }
+        }
+        return vj_indexes;
```
Comment (pbench) [inline source/ptreferential/ptreferential_utils.cpp]:
```suggestion
        Indexes impact_indexes, temp;
        for (type::VehicleJourney* vj : vjs) {
            if (indexes.find(vj->physical_mode->idx) != indexes.cend()) {
                // add all impacts that concern this vehicle_journey to impact_indexes
                temp = vj->get(Type_e::Impact, *(data.pt_data.get()));
                impact_indexes.insert(temp.begin(), temp.end());
            }
        }
        return vj_indexes;
```

File: source/ptreferential/ptreferential_utils.cpp [hunk 2]
```diff
@@ -69,6 +69,24 @@ Indexes get_intersection(const Indexes& idxs1, const Indexes& idxs2) {
 }
 
 Indexes get_corresponding(Indexes indexes, Type_e from, const Type_e to, const Data& data) {
+    // Exceptional case: if from = PhysicalMode and to = Impact
+    // 1. Get all vehicle_journeys impacted
+    // 2. Keep only vehicle_journeys with physical_mode in the parameter
+
+    if (from == Type_e::PhysicalMode && to == Type_e::Impact) {
+        auto vj_idxs = get_indexes_by_impacts(Type_e::VehicleJourney, data, false);
```
Comment (pbench) [inline source/ptreferential/ptreferential_utils.cpp:76]:
```suggestion
    if (from == Type_e::PhysicalMode && to == Type_e::Impact) {
        // recover the indexes of all vehicle_journeys that have an impact on them
        auto vj_idxs = get_indexes_by_impacts(Type_e::VehicleJourney, data, false);
```

File: source/jormungandr/tests/kirin_realtime_tests.py [hunk 3]
```diff
@@ -2446,6 +2446,10 @@ def test_update_trip_with_physical_mode(self):
         assert len(pt_response['physical_modes']) == 1
         assert pt_response['physical_modes'][0]['name'] == 'Tramway'
 
+        # Check disruptions with physical_modes
+        disruptions_from_mode = self.query_region('physical_modes/physical_mode:0x0/disruptions')
```
Comment (benoit-bst) [inline source/jormungandr/tests/kirin_realtime_tests.py:2450]: yes, it is enough to test. And for the perfomance we have the real test

**Review by benoit-bst — APPROVED** [review]

**Review by xlqian — APPROVED** [review]

**Review by pbench — APPROVED** [review]

**Comment by kadhikari** [comment]
Tested with kirin realtime data PIV de sncf on DEV with @pbench

**Pull request merged**